core/src/main/java/edu/wpi/grip/core/http/GripServer.java

```diff
@@ -0,0 +1,215 @@
+import org.eclipse.jetty.server.handler.HandlerCollection;
+
+/**
+ * An internal HTTP server.
+ */
+@Singleton
+public class GripServer {
+
+    /**
+     * Factory for creating a new Jetty server.
+     */
+    private final JettyServerFactory serverFactory;
+
+    /**
+     * The port that this server is running on.
+     */
+    private int port;
+
+    /**
+     * The internal Jetty server that actually handles all server operations.
+     */
+    private Server server;
+
+    /**
+     * A collection of Jetty handlers that can be added to or removed during runtime.
+     */
+    private final HandlerCollection handlers = new HandlerCollection(true);
+
+    /**
+     * The current lifecycle state of the server.
+     */
+    private State state = State.PRE_RUN;
+
+    private enum State {
+        PRE_RUN,
+        RUNNING,
+        STOPPED
+    }
+
+    /**
+     * The root path for all GRIP-related HTTP activity.
+     */
+    public static final String ROOT_PATH = "/GRIP";
+
+    /**
+     * The root path for uploading data to the server.
+     */
+    public static final String UPLOAD_PATH = ROOT_PATH + "/upload";
+
+    /**
+     * The path for uploading images. To upload an image, post an HTTP event to
+     * {@code /GRIP/upload/image}, with the image bytes as the data.
+     */
+    public static final String IMAGE_UPLOAD_PATH = UPLOAD_PATH + "/image";
+
+    /**
+     * The path for setting which pipeline to run. To set the pipeline, post an
+     * HTTP event to {@code /GRIP/upload/pipeline},
+     * with the content of the pipeline save file as the data.
+     */
+    public static final String PIPELINE_UPLOAD_PATH = UPLOAD_PATH + "/pipeline";
+
+    /**
+     * The path for requesting data. Data will be returned as a json-formatted
+     * map of the outputs of all requested data sets.
+     * <p>
+     * For example, performing a {@code GET} request on the path
+     * {@code /GRIP/data?foo&bar} will return a map such as
+     * <br>
+     * <code><pre>
+     * {
+     *   'foo': {
+     *        // data
+     *      },
+     *   'bar':
+     *     {
+     *       // data
+     *     }
+     * }
+     * </pre></code>
+     */
+    public static final String DATA_PATH = ROOT_PATH + "/data";
+
+    public interface JettyServerFactory {
+        Server create(int port);
+    }
+
+    public static class JettyServerFactoryImpl implements JettyServerFactory {
+
+        @Override
+        public Server create(int port) {
+            return new Server(port);
+        }
+    }
+
+    @Inject
+    GripServer(JettyServerFactory serverFactory, SettingsProvider settingsProvider) {
+        this.port = settingsProvider.getProjectSettings().getServerPort();
+        this.serverFactory = serverFactory;
+        this.server = serverFactory.create(port);
+        this.server.setHandler(handlers);
+        handlers.addHandler(new NoContextHandler());
+
+        Runtime.getRuntime().addShutdownHook(new Thread(this::stop));
```

Comment:
Basically tells Jetty to stop handling new requests and finish up any currently running handlers. Otherwise it'll keep responding to stuff until the server thread gets killed by the JVM.